### PR TITLE
#258 normalize triple slash comments xml

### PR DESCRIPTION
**Pull request by ansyral**
@vwxyzh @hellosnow @superyyrrzz @qinezh @chenkennt

### CODE REVIEW

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/Parsers/TripleSlashCommentModel.cs [hunk 1]
```diff
@@ -384,37 +379,46 @@ private string GetXmlValue(XPathNavigator node)
...
         }
 
-        private static string NormalizeLine(string line, int column)
+        /// <summary>
+        /// Split xml to lines. Trim meaningless whitespaces and newline.
+        /// if a line starts with xml node, all leading whitespaces would be trimmed
+        /// otherwise text node start position always aligns with the start position of its parent line(the last previous line that starts with xml node)
+        /// the newline character after <node> and before </node> would be removed
+        /// </summary>
+        /// <param name="xml"></param>
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/Parsers/TripleSlashCommentModel.cs]: add description for parentIndex?

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/Parsers/TripleSlashCommentModel.cs [hunk 2]
```diff
@@ -383,38 +378,46 @@ private string GetXmlValue(XPathNavigator node)
...
+        /// <returns></returns>
+        private static string NormalizeXml(string xml, int parentIndex)
         {
-            int trimIndex = Math.Min(column, GetNonWhitespaceIndex(line));
+            var lines = new Regex(@"\r?\n").Split(xml);
```
Comment (qinezh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/Parsers/TripleSlashCommentModel.cs]: change to static field with RegexOptions.Compiled just like `CommentIdRegex`?

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/Parsers/TripleSlashCommentModel.cs [hunk 3]
```diff
@@ -383,38 +378,46 @@ private string GetXmlValue(XPathNavigator node)
...
+                if (string.IsNullOrWhiteSpace(line))
+                {
+                    normalized.Add("\n");
+                    continue;
+                }
+                if (Regex.IsMatch(line, @"^\s*<\s*/") && normalized.LastOrDefault() == "\n")
+                {
+                    normalized.RemoveAt(normalized.Count - 1);
+                }
+                int index = line.TakeWhile(char.IsWhiteSpace).Count();
```
Comment (vwxyzh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/Parsers/TripleSlashCommentModel.cs]: count {TAB} as 4 {SPACE}?

Comment (ansyral) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/Parsers/TripleSlashCommentModel.cs]: seems that the input couldn't start with tab?

**Comment by vwxyzh** [comment]
👍

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/Parsers/TripleSlashCommentModel.cs [hunk 4]
```diff
@@ -22,6 +22,8 @@ public class TripleSlashCommentModel
     {
         private const string idSelector = @"((?![0-9])[\w_])+[\w\(\)\.\{\}\[\]\|\*\^~#@!`,_<>:]*";
         private static Regex CommentIdRegex = new Regex(@"^(?<type>N|T|M|P|F|E):(?<id>" + idSelector + ")$", RegexOptions.Compiled);
+        private static Regex LineBreakRegex = new Regex(@"\r?\n", RegexOptions.Compiled);
+        private static Regex codeElementRegex = new Regex(@"<code[^>]*>([\s\S]*?)</code>", RegexOptions.Compiled);
```
Comment (qinezh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/Parsers/TripleSlashCommentModel.cs]: consistent naming style with above private static field?

Comment (qinezh) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/Parsers/TripleSlashCommentModel.cs]: match any character?  `.` could be better in this case

Comment (ansyral) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/Parsers/TripleSlashCommentModel.cs]: `.` seems not to match newline character?

File: src/Microsoft.DocAsCode.Metadata.ManagedReference/Parsers/TripleSlashCommentModel.cs [hunk 5]
```diff
@@ -383,38 +380,43 @@ private string GetXmlValue(XPathNavigator node)
...
-            for (var i = 1; i < lines.Length; i++)
-            {
-                yield return NormalizeLine(lines[i], column);
-            }
+            return WebUtility.HtmlDecode(NormalizeXml(node.InnerXml, column));
         }
 
-        private static string NormalizeLine(string line, int column)
+        /// <summary>
+        /// Split xml to lines. Trim meaningless whitespaces.
```
Comment (superyyrrzz) [inline src/Microsoft.DocAsCode.Metadata.ManagedReference/Parsers/TripleSlashCommentModel.cs]: split ... to => split ... into

**Comment by superyyrrzz** [comment]
:shipit:

**Pull request merged**